### DISCUSSION TITLE
How to use NATS over serial?

### DISCUSSION THREAD

**Discussion by maxwang7** [nats-io/nats.rs#1328] [Q&A]
Hi! I'm using NATS / async-nats (rust) for an embedded application on an ESP32 connected to a Raspberry Pi, and would like to extend NATS over serial so that my ESP32 is able to directly talk over the NATS network. Is there a way to use async-nats on a serial port? Or is there an idiomatic way of creating a networking setup where the ESP32 is able to receive TCP messages? If not, is there a recommended way of running TCP over serial so that embedded devices are able to connect to the NATS network? So far I have a TCP to serial connection setup using socat on the Raspberry Pi but struggling to get the ESP32 being able to read from the serial port using NATS. More details below on setup:

Nodes:
- Cloud box - runs nats server as a hub node
- Raspberry Pi - runs nats server as a leaf node, connected over the internet to the cloud node
- ESP32 - connected to Raspberry Pi over USB serial, using esp-idf framework so able to use TCP/IP stack

Thanks in advance!

**Comment by Jarema** [comment]
I haven't try it yet, but I think that using ppp or slip might be the easiest way to solve the problem.
Maybe @paolobarbolini has more insights here.

**Reply by paolobarbolini** [reply]
I haven't had to do it yet but the ppp/slip route seems much more reliable than piping a connection based protocol over serial and expecting everything to work right.

**Reply by maxwang7** [reply]
Thanks for the tip! For others to reference, I found the espressif has the eppp_link component: https://components.espressif.com/components/espressif/eppp_link/versions/0.2.0. Has PPPoS support. Will give that a shot.
